firedrake/interpolation.py

```diff
@@ -286,28 +359,133 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
         parloop_args.append(tensor(access, V.cell_node_map()))
     else:
         assert access == op2.WRITE  # Other access descriptors not done for Matrices.
-        parloop_args.append(tensor(op2.WRITE, (V.cell_node_map(),
-                                               arguments[0].function_space().cell_node_map())))
+        rows_map = V.cell_node_map()
+        columns_map = arguments[0].function_space().cell_node_map()
+        if trans_mesh:
+            # We have to map to the source mesh cells before mapping to nodes
+            columns_map = compose_map_and_cache(target_mesh_source_mesh_map,
+                                                columns_map)
+        parloop_args.append(tensor(op2.WRITE, (rows_map, columns_map)))
     if oriented:
-        co = mesh.cell_orientations()
+        co = target_mesh.cell_orientations()
         parloop_args.append(co.dat(op2.READ, co.cell_node_map()))
     if needs_cell_sizes:
-        cs = mesh.cell_sizes
+        cs = target_mesh.cell_sizes
         parloop_args.append(cs.dat(op2.READ, cs.cell_node_map()))
     for coefficient in coefficients:
-        m_ = coefficient.cell_node_map()
-        parloop_args.append(coefficient.dat(op2.READ, m_))
-
-    for o in coefficients:
-        domain = o.ufl_domain()
-        if domain is not None and domain.topology != mesh.topology:
-            raise NotImplementedError("Interpolation onto another mesh not supported.")
+        if coefficient.ufl_domain() == source_mesh and trans_mesh:
+            m_ = compose_map_and_cache(target_mesh_source_mesh_map, coefficient.cell_node_map())
+            parloop_args.append(coefficient.dat(op2.READ, m_))
+        else:
+            m_ = coefficient.cell_node_map()
+            parloop_args.append(coefficient.dat(op2.READ, m_))
```

Comment:
(Suggested) The following would be slightly more intuitive:
```
for coefficient in coefficients:
    coeff_mesh = coefficient.ufl_domain()
    if coeff_mesh is target_mesh or not coeff_mesh:
        m_ = coefficient.cell_node_map()
    elif coeff_mesh is source_mesh:
        m_ = compose_map_and_cache(target_mesh.cell_parent_cell_map, coefficient.cell_node_map())
    else:
        raise ...
    parloop_args.append(coefficient.dat(op2.READ, m_))
```

Reply:
Agreed

Reply:
Done